sam.c

```diff
@@ -2398,87 +2564,67 @@ static int sam_parse_B_vals(char type, uint32_t n, char *in, char **end,
         return -1;
     }
 
-    // Ensure space for type + values
-    bytes = (size_t) n * (size_t) size;
-    if (bytes / size != n
-        || possibly_expand_bam_data(b, bytes + 2 + sizeof(uint32_t))) {
-        hts_log_error("Out of memory");
-        return -1;
+    // Ensure space for type + values.
+    // The first pass through here we don't know the number of entries and
+    // nalloc == 0.  We start with a small working set and then parse the
+    // data, growing as needed.
+    //
+    // If we have a second pass through we do know the number of entries
+    // and nalloc is already known.  We have no need to expand the bam data.
+    if (!nalloc) {
+         nalloc=7;
+
+         bytes = (size_t) nalloc * (size_t) size;
```

Comment:
This was correct before.  If recursing, `size` may have changed from the value used in the original allocation.  If it has, you need to get some extra space.  `possibly_expand_bam_data()` will not do anything if the space already available is adequate.

You can test this using a tag like `bb:B:c,1000000,,,,,,,,,,,,,,,,,,,,,,,,,,`, making it long enough to need a reallocation.

Reply:
Ah yes I forgot about the growth case.  However it's still not correct before.  Eg it could grow by Nx1 (N "c" records), and then grow by a further Nx4 when recoding to int.  It probably doesn't matter much though as growing too much isn't a major concern given the number of records it affects *should* be tiny and we can only do 1 extra pass.

Reply:
Plus, tbh, we should probably have always just flat out rejected such data as an error.  I don't know why we're bending over backwards to fix broken SAM records really.

Reply:
It was correct, because `b->l_data` gets reset, and `possibly_expand_bam_data()` measures the expansion requested from there.  If `size` hasn't changed, it will do nothing.

Reply:
Ah I see.  Cryptic!  I hadn't noticed the reset of `b->l_data` in the recursion section.  I do question the whole ethos of fixing up broken data instead of simply rejecting it as it just encourages people to write bad code given we correct their mistakes for them, but we are where we are so reverted that change back again.